src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,320 @@
+//-*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include <seastar/core/file.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/reactor.hh>
+#include <linux/nvme_ioctl.h>
+#include <libaio.h>
+
+#include "crimson/osd/exceptions.h"
+#include "crimson/common/layout.h"
+
+namespace ceph {
+  namespace buffer {
+    class bufferptr;
+  }
+}
+
+namespace crimson::os::seastore::nvme_device {
+
+/*
+ * NVMe protocol structures (nvme_XX, identify_XX)
+ *
+ * All structures relative to NVMe protocol are following NVMe protocol v1.4
+ * (latest). NVMe is protocol for fast interfacing between user and SSD device.
+ * We selectively adopted features among various NVMe features to ease
+ * implementation. And also, NVMeBlockDevice provides generic command submission
+ * APIs for IO and Admin commands. Please use pass_through_io() and pass_admin()
+ * to do it.
+ *
+ * For more information about NVMe protocol, refer https://nvmexpress.org/
+ */
+
+struct nvme_identify_command_t {
+  uint32_t common_dw[10];
+  uint32_t cns : 8;
+  uint32_t reserved : 8;
+  uint32_t cntroller_id : 16;
+
+  static const uint8_t CNS_NAMESPACE = 0x00;
+  static const uint8_t CNS_CONTROLLER = 0x01;
+};
+
+struct nvme_admin_command_t {
+  union
+  {
+    nvme_passthru_cmd common_cmd;
+    nvme_identify_command_t identify_cmd;
+  };
+
+  static const uint8_t OPCODE_IDENTIFY = 0x06;
+};
+
+struct nvme_version_t {
+  uint32_t major_ver : 16;
+  uint32_t minor_ver : 8;
+  uint32_t tertiary_ver : 8;
+};
+
+struct admin_command_support_t {
+  uint16_t unused : 5;
+  uint16_t support_directives : 1;
+  uint16_t unused2 : 10;
+};
+
+struct identify_controller_data_t {
+  union
+  {
+    struct
+    {
+      uint8_t raw[1024];
+    };
+    struct
+    {
+      uint8_t unused[80];
+      nvme_version_t version;
+      uint8_t unused2[172];
+      admin_command_support_t oacs;
+    };
+  };
+};
+
+struct identify_namespace_data_t {
+  union
+  {
+    struct
+    {
+      uint8_t raw[4096];
+    };
+    struct
+    {
+      uint8_t unused[64];
+      uint16_t npwg;
+      uint16_t npwa;
+    };
+  };
+};
```

Comment:
These bits are unused at the moment, right?

Reply:
Right. Removed.